avocado/core/runner.py

```diff
@@ -485,36 +485,57 @@ def sigtstp_handler(signum, frame):     # pylint: disable=W0613
         return True
 
     @staticmethod
-    def _iter_variants(template, variants):
+    def _template_to_factory(template, variant):
         """
-        Iterate through variants and set the params/variants accordingly.
+        Applies test params from variant to the test template
 
-        :param template: test template
-        :param variants: the Mux object containing the variants
-        :return: Yields tuple(test_factory including params, variant id)
-        :raises ValueError: When variant and template declare params.
+        :param template: a test template
+        :param variant: variant to be applied
+        :return: tuple(new_test_factory, applied_variant)
         """
-        for variant in variants.itertests():
-            params = (variant.get("variant"), variant.get("mux_path"))
-            if params:
-                if "params" in template[1]:
-                    msg = ("Unable to use test variants %s, params are already"
-                           " present in test factory: %s"
-                           % (template[0], template[1]))
-                    raise ValueError(msg)
-                factory = [template[0], template[1].copy()]
-                factory[1]["params"] = params
-            else:
-                factory = template
-            yield factory, variant
+        params = variant.get("variant"), variant.get("mux_path")
+        if params:
+            if "params" in template[1]:
+                msg = ("Unable to use test variants %s, params are already"
+                       " present in test factory: %s"
+                       % (template[0], template[1]))
+                raise ValueError(msg)
+            factory = [template[0], template[1].copy()]
+            factory[1]["params"] = params
+        else:
+            factory = template
+        return factory, variant
 
-    def run_suite(self, test_suite, variants, timeout=0, replay_map=None):
+    def _iter_suite(self, test_suite, variants, suite_order):
+        """
+        Iterates through test_suite and variants in defined order
+
+        :param test_suite: a list of tests to run
+        :param variants: a varianter object to produce test params
+        :param suite_order: way of iterating through tests/variants
+        :return: generator yielding tuple(test_factory, variant)
+        """
+        if suite_order == "tests-per-variant":
+            return (self._template_to_factory(template, variant)
+                    for variant in variants.itertests()
+                    for template in test_suite)
+        else:
```

Comment:
well I wasn't sure about that so I went with "default" in case nothing is provided, but sure, I can raise exception...